### ISSUE TITLE
TensorRT 8 TopK output accuracy pretty low

### ISSUE THREAD

**Issue by jackgao0323** [NVIDIA/TensorRT#2259]
## Description

I using polygraphy and found my output is exceeds tolerance and I found it starts from TopK layer.
![image](https://user-images.githubusercontent.com/17614632/185579826-19bc1501-2e61-4df2-abd9-f5d23f18d10a.png)

I make a model only using one TopK layer and found the output is not correct.

![image](https://user-images.githubusercontent.com/17614632/185581419-0a7ec6ea-9469-457d-9c26-e489dd02997f.png)

![image](https://user-images.githubusercontent.com/17614632/185581533-9fd1363f-f3b4-4549-8ace-5867848cd43b.png)

I have using TRT python and onnxruntime to compare the cosine similarity and it's pretty low.

![image](https://user-images.githubusercontent.com/17614632/185581857-324495c5-3792-41a3-9ca2-3dbeeedb0c32.png)

## Environment

**TensorRT Version**: 8.4.2.4
**NVIDIA GPU**: RTX 3060ti
**NVIDIA Driver Version**: 515.65.01
**CUDA Version**: 11.7
**CUDNN Version**: 8.4.0
**Operating System**: Linux
**Python Version (if applicable)**: 3.9
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**: 

## Relevant Files

[Compress.tar.gz](https://github.com/NVIDIA/TensorRT/files/9380625/Compress.tar.gz)

## Steps To Reproduce

polygraphy run ./TopK.onnx --trt --onnxrt --pool-limit workspace:2G

**Comment by zerollzeng** [comment]
Hi @jackgao0323 This is indeed a TRT onnx parser bug. the reason why you see this bug is the parser always parse the topK as "top max k", but in your case, it's 'top small k"
![image](https://user-images.githubusercontent.com/38289304/185741219-5ef4c8a6-2ed5-4d86-a8c7-2892b059025a.png)

to quickly WAR this, you can modify the https://github.com/onnx/onnx-tensorrt/blob/bf5f48f2ffcdd4734f2a07c08571313180adeffd/builtin_op_importers.cpp#L4460 as below
```
-    nvinfer1::ITopKLayer* layer = ctx->network()->addTopK(*tensorPtr, nvinfer1::TopKOperation::kMAX, k, axisMask);
+    // Default is top max k.
+    auto operation = nvinfer1::TopKOperation::kMAX;
+    if (ctx->getOpsetVersion() >= 11)
+    {
+        ASSERT( (attrs.count("largest")) && "Attribute largest is missing.", ErrorCode::kINVALID_NODE);
+        int32_t largest = attrs.get<int>("largest");
+        if (largest == 0)
+        {
+            operation = nvinfer1::TopKOperation::kMIN;
+        }
+        // TRT only produce sorted output, but we can still check it to match onnx specs.
+        ASSERT( (attrs.count("sorted")) && "Attribute sorted is missing.", ErrorCode::kINVALID_NODE);
+    }
+    nvinfer1::ITopKLayer* layer = ctx->network()->addTopK(*tensorPtr, operation, k, axisMask);
```

**Comment by zerollzeng** [comment]
Sorry for the inconvenience, I will file an MR to fix this internally. cc @kevinch-nv

**Comment by zerollzeng** [comment]
after fix:
```
[I] Accuracy Comparison | trt-runner-N0-08/20/22-03:07:10 vs. onnxrt-runner-N0-08/20/22-03:07:10
[I]     Comparing Output: 'patch_scores' (dtype=float32, shape=(784, 9)) with 'patch_scores' (dtype=float32, shape=(784, 9))
[I]     Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[I]         trt-runner-N0-08/20/22-03:07:10: patch_scores | Stats: mean=0.00028983, std-dev=0.00019977, var=3.9908e-08, median=0.00025827, min=3.0078e-07 at (42, 0), max=0.0012891 at (471, 8), avg-magnitude=0.00028983
[I]         onnxrt-runner-N0-08/20/22-03:07:10: patch_scores | Stats: mean=0.00028983, std-dev=0.00019977, var=3.9908e-08, median=0.00025827, min=3.0078e-07 at (42, 0), max=0.0012891 at (471, 8), avg-magnitude=0.00028983
[I]         Error Metrics: patch_scores
[I]             Minimum Required Tolerance: elemwise error | [abs=0] OR [rel=0] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0), max=0 at (0, 0), avg-magnitude=0
[I]             Relative Difference | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0), max=0 at (0, 0), avg-magnitude=0
[I]         PASSED | Difference is within tolerance (rel=1e-05, abs=1e-05)
[I]     Comparing Output: 'indices' (dtype=int32, shape=(784, 9)) with 'indices' (dtype=int64, shape=(784, 9))
[I]     Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[I]         trt-runner-N0-08/20/22-03:07:10: indices | Stats: mean=8625.8, std-dev=4994.9, var=2.4949e+07, median=8539, min=2 at (0, 3), max=17240 at (417, 0), avg-magnitude=8625.8
[I]         onnxrt-runner-N0-08/20/22-03:07:10: indices | Stats: mean=8625.8, std-dev=4994.9, var=2.4949e+07, median=8539, min=2 at (0, 3), max=17240 at (417, 0), avg-magnitude=8625.8
[I]         Error Metrics: indices
[I]             Minimum Required Tolerance: elemwise error | [abs=0] OR [rel=0] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0), max=0 at (0, 0), avg-magnitude=0
[I]             Relative Difference | Stats: mean=0, std-dev=0, var=0, median=0, min=0 at (0, 0), max=0 at (0, 0), avg-magnitude=0
[I]         PASSED | Difference is within tolerance (rel=1e-05, abs=1e-05)
[I]     PASSED | All outputs matched | Outputs: ['patch_scores', 'indices']
[I] PASSED | Command: /home/zeroz/.local/bin/polygraphy run ./TopK.onnx --trt --onnxrt

```

**Comment by jackgao0323** [comment]
> Hi @jackgao0323 This is indeed a TRT onnx parser bug. the reason why you see this bug is the parser always parse the topK as "top max k", but in your case, it's 'top small k" ![image](https://user-images.githubusercontent.com/38289304/185741219-5ef4c8a6-2ed5-4d86-a8c7-2892b059025a.png)
> 
> to quickly WAR this, you can modify the https://github.com/onnx/onnx-tensorrt/blob/bf5f48f2ffcdd4734f2a07c08571313180adeffd/builtin_op_importers.cpp#L4460 as below and recompile the parser.
> 
> ```
> -    nvinfer1::ITopKLayer* layer = ctx->network()->addTopK(*tensorPtr, nvinfer1::TopKOperation::kMAX, k, axisMask);
> +    // Default is top max k.
> +    auto operation = nvinfer1::TopKOperation::kMAX;
> +    if (ctx->getOpsetVersion() >= 11)
> +    {
> +        ASSERT( (attrs.count("largest")) && "Attribute largest is missing.", ErrorCode::kINVALID_NODE);
> +        int32_t largest = attrs.get<int>("largest");
> +        if (largest == 0)
> +        {
> +            operation = nvinfer1::TopKOperation::kMIN;
> +        }
> +        // TRT only produce sorted output, but we can still check it to match onnx specs.
> +        ASSERT( (attrs.count("sorted")) && "Attribute sorted is missing.", ErrorCode::kINVALID_NODE);
> +    }
> +    nvinfer1::ITopKLayer* layer = ctx->network()->addTopK(*tensorPtr, operation, k, axisMask);
> ```

Hi @zerollzeng ,
Thanks for your reply. Will it fixed in the next version?

BR,
Jack Gao

**Comment by zerollzeng** [comment]
Thanks for your reply. Will it fixed in the next version?
-> Yes, should be fixed in the TRT 8.5

**Comment by jackgao0323** [comment]
@zerollzeng Is this bug fixed in TRT 8.5GA update1? Can you help me check?

**Comment by jackgao0323** [comment]
@zerollzeng Any update of this issue?

**Comment by zerollzeng** [comment]
Yes it's fixed in 8.5GA.
